lib/PuppeteerSharp.Tests/Target/Tests.cs

```diff
@@ -0,0 +1,183 @@
+            Assert.NotNull(await otherPage.GetElementAsync("body"));
+
+            var allPages = await Browser.Pages();
+            Assert.Contains(Page, allPages);
+            Assert.Contains(otherPage, allPages);
+
+            var closePageTaskCompletion = new TaskCompletionSource<PuppeteerSharp.Page>();
+            async void TargetDestroyedEventHandler(object sender, TargetChangedArgs e)
+            {
+                closePageTaskCompletion.SetResult(await e.Target.Page());
+                Browser.TargetDestroyed -= TargetDestroyedEventHandler;
+            }
+            Browser.TargetDestroyed += TargetDestroyedEventHandler;
+            await otherPage.CloseAsync();
+            Assert.Equal(otherPage, await closePageTaskCompletion.Task);
+
+            allPages = await Task.WhenAll(Browser.Targets().Select(target => target.Page()));
+            Assert.Contains(Page, allPages);
+            Assert.DoesNotContain(otherPage, allPages);
+        }
+
+        [Fact]
+        public async Task ShouldReportWhenAServiceWorkerIsCreatedAndDestroyed()
+        {
+            await Page.GoToAsync(TestConstants.EmptyPage);
+            var createdTargetTaskCompletion = new TaskCompletionSource<PuppeteerSharp.Target>();
+            void TargetCreatedEventHandler(object sender, TargetChangedArgs e)
+            {
+                createdTargetTaskCompletion.SetResult(e.Target);
+                Browser.TargetCreated -= TargetCreatedEventHandler;
+            }
+            Browser.TargetCreated += TargetCreatedEventHandler;
+            var registration = await Page.EvaluateExpressionHandle("navigator.serviceWorker.register('sw.js')");
+
+            var createdTarget = await createdTargetTaskCompletion.Task;
+            Assert.Equal("service_worker", createdTarget.Type);
+            Assert.Equal(TestConstants.ServerUrl + "/sw.js", createdTarget.Url);
+
+            var targetDestroyedTaskCompletion = new TaskCompletionSource<PuppeteerSharp.Target>();
+            void TargetDestroyedEventHandler(object sender, TargetChangedArgs e)
+            {
+                targetDestroyedTaskCompletion.SetResult(e.Target);
+                Browser.TargetDestroyed -= TargetDestroyedEventHandler;
+            }
+            Browser.TargetDestroyed += TargetDestroyedEventHandler;
+            await Page.EvaluateFunctionAsync("registration => registration.unregister()", registration);
+            Assert.Equal(createdTarget, await targetDestroyedTaskCompletion.Task);
+        }
+
+        [Fact]
+        public async Task ShouldReportWhenATargetUrlChanges()
+        {
+            await Page.GoToAsync(TestConstants.EmptyPage);
+
+            var changedTargetTaskCompletion = new TaskCompletionSource<PuppeteerSharp.Target>();
+            void ChangedTargetEventHandler(object sender, TargetChangedArgs e)
+            {
+                changedTargetTaskCompletion.SetResult(e.Target);
+                Browser.TargetChanged -= ChangedTargetEventHandler;
```

Comment:
This is because in NodeJS is `once` right?

Reply:
exactly